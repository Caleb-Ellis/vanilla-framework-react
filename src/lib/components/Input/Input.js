import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

function inputTag(type) {
  if (type === 'select') {
    return 'select';
  } else if (type === 'textarea') {
    return 'textarea';
  }
  return 'input';
}

const Input = (props) => {
  const {
    className, hasValidation, type, ...otherProps
  } = props;

  const classNames = getClassName({
    [className]: className,
    'p-form-validation__input': hasValidation,
  }) || undefined;

  const Tag = inputTag(type);

  return (
    <Tag
      className={classNames}
      type={type}
      {...otherProps}
    />
  );
};

Input.defaultProps = {
  className: undefined,
  type: 'text',
  hasValidation: false,
};

Input.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  type: PropTypes.oneOf([
    'text', 'password', 'file', 'checkbox', 'radio', 'select', 'textarea', 'date',
    'hidden', 'month', 'time', 'week', 'color', 'number', 'email', 'url', 'tel',
  ]),
  hasValidation: PropTypes.bool,
};

Input.displayName = 'Input';

export default Input;
