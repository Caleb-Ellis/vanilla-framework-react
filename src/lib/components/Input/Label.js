import React from 'react';
import PropTypes from 'prop-types';
import getClassName from '../../utils/getClassName';

const Label = (props) => {
  const {
    children, htmlFor, className, ...otherProps
  } = props;

  const classNames = getClassName({
    [className]: className,
    'p-form__label': true,
  }) || undefined;

  return (
    <label className={classNames} htmlFor={htmlFor} {...otherProps}>{children}</label> // eslint-disable-line
  );
};

Label.defaultProps = {
  children: null,
  className: undefined,
};

Label.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  htmlFor: PropTypes.string.isRequired,
};

Label.displayName = 'Label';

export default Label;
