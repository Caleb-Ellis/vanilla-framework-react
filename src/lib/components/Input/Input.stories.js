import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import Input from './Input';
import Label from './Label';
import Button from '../Button/Button';
import Form from '../Form/Form';

const inputTypes = [
  'text', 'password', 'file', 'hidden', 'date', 'month',
  'time', 'week', 'number', 'email', 'url', 'tel',
];

const exampleSubmit = (event) => {
  event.preventDefault();
  alert('You submitted the form!'); // eslint-disable-line no-alert
};

storiesOf('Input', module)
  .add('Standard inputs',
    withInfo('Vanilla <Input> components support most HTML5 input types: text, password, file, hidden, date, month, time, week, number, email, url and tel. Note that a dummy submit function has been provided to prevent the page refreshing.')(() => (
      <Form onSubmit={exampleSubmit}>
        <Label htmlFor="textInput1">{text('Label', 'HTML5 input')}</Label>
        <Input
          type={select('Input type', inputTypes, 'text')}
          id="textInput1"
          placeholder={text('Placeholder', 'Placeholder')}
          maxLength={text('Max length', '')}
          required={boolean('Required', false)}
          disabled={boolean('Disabled', false)}
        />
        <Button neutral type="submit">Submit</Button>
      </Form>),
    ),
  )

  .add('Checkbox and Radio',
    withInfo('Use checkboxes and radio buttons to select one or more options. Note that <Input> components are placed before <Label> components for checkboxes and radio buttons.')(() => (
      <Form onSubmit={exampleSubmit}>
        <Input
          type={select('Type1', ['checkbox', 'radio'], 'checkbox')}
          id="input1"
          required={boolean('Required1', false)}
          disabled={boolean('Disabled1', false)}
          name="inputs"
        />
        <Label htmlFor="input1">{text('Label1', 'Input 1')}</Label>
        <Input
          type={select('Type2', ['checkbox', 'radio'], 'checkbox')}
          id="input2"
          required={boolean('Required2', false)}
          disabled={boolean('Disabled2', false)}
          name="inputs"
        />
        <Label htmlFor="input2">{text('Label2', 'Input 2')}</Label>
        <Button neutral type="submit">Submit</Button>
      </Form>),
    ),
  )

  .add('Textarea',
    withInfo('An <Input> with type="textarea" defines a multi-line text input control.')(() => (
      <Form onSubmit={exampleSubmit}>
        <Label htmlFor="textareaInput1">Textarea</Label>
        <Input
          type="textarea"
          id="textareaInput1"
          placeholder={text('Placeholder', 'Placeholder')}
          rows={text('Rows', '3')}
          required={boolean('Required', false)}
          disabled={boolean('Disabled', false)}
        />
        <Button neutral type="submit">Submit</Button>
      </Form>),
    ),
  )

  .add('Select',
    withInfo('An <Input> with type="select" creates a drop-down list.')(() => (
      <Form onSubmit={exampleSubmit}>
        <Label htmlFor="selectInput1">Select</Label>
        <Input
          type="select"
          id="selectInput1"
          name="selectExample1"
          required={boolean('Required', false)}
          disabled={boolean('Disabled', false)}
          defaultValue=""
          multiple={boolean('Multiple', false)}
        >
          <option value="" disabled>Select an option</option>
          <option value="1">{text('Option 1', 'Option 1')}</option>
          <option value="2">{text('Option 2', 'Option 2')}</option>
          <option value="3">{text('Option 3', 'Option 3')}</option>
        </Input>
        <Button neutral type="submit">Submit</Button>
      </Form>),
    ),
  );
