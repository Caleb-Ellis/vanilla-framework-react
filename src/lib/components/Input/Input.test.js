import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Input from './Input';
import Label from './Label';

describe('<Input>', () => {
  it('should match the default snapshot', () => {
    const input = ReactTestRenderer.create(
      <Input id="input" />);
    const json = input.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render an input tag by default', () => {
    const input = shallow(
      <Input id="input" />,
    );

    expect(input.type()).toBe('input');
  });

  it('should render a select tag if type is "select"', () => {
    const input = shallow(
      <Input id="input" type="select" />,
    );

    expect(input.type()).toBe('select');
  });

  it('should render a textarea tag if type is "textarea"', () => {
    const input = shallow(
      <Input id="input" type="textarea" />,
    );

    expect(input.type()).toBe('textarea');
  });

  it('should render an input tag if type is not a special case', () => {
    const input = shallow(
      <Input id="input" type="password" />,
    );

    expect(input.type()).toBe('input');
  });

  it('should have appropriate validation class if hasValidation prop is true', () => {
    const validationClass = 'p-form-validation__input';
    const input = shallow(
      <Input id="input" hasValidation />,
    );

    expect(input.hasClass(validationClass)).toBe(true);
  });

  it('should accept custom classnames', () => {
    const input = shallow(
      <Input id="input" className="custom-class" />,
    );

    expect(input.hasClass('custom-class')).toBe(true);
  });
});

describe('<Label>', () => {
  it('should match the default snapshot', () => {
    const label = ReactTestRenderer.create(
      <div>
        <Input id="input" />
        <Label htmlFor="input" />
      </div>);
    const json = label.toJSON();
    expect(json).toMatchSnapshot();
  });
});
