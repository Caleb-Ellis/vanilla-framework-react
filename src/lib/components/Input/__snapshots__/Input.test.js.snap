// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input> should match the default snapshot 1`] = `
<input
  className={undefined}
  id="input"
  type="text"
/>
`;

exports[`<Label> should match the default snapshot 1`] = `
<div>
  <input
    className={undefined}
    id="input"
    type="text"
  />
  <label
    className="p-form__label"
    htmlFor="input"
  />
</div>
`;
